util/perl/OpenSSL/OID.pm

```diff
@@ -0,0 +1,365 @@
+# Copyright 2016-2019 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+# Author note: this is originally RL::ASN1::OID,
+# repurposed by the author for OpenSSL use.
+
+package OpenSSL::OID;
+
+use 5.10.0;
+use strict;
+use warnings;
+use Carp;
+
+use Exporter;
+use vars qw($VERSION @ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
+@ISA = qw(Exporter);
+@EXPORT = qw(parse_oid encode_oid register_oid
+             registered_oid_arcs registered_oid_leaves);
+@EXPORT_OK = qw(encode_oid_nums);
+
+use List::Util;
+
+=head1 NAME
+
+OpenSSL::OID - an OBJECT IDENTIFIER parser / encoder
+
+=head1 VERSION
+
+Version 0.1
+
+=cut
+
+our $VERSION = '0.1';
+
+
+=head1 SYNOPSIS
+
+    use OpenSSL::OID;
+
+    # This gives the array ( 1 2 840 113549 1 1 )
+    my @nums = parse_oid('{ pkcs-1 1 }');
+
+    # This gives the array of DER encoded bytes for the OID, i.e.
+    # ( 42, 134, 72, 134, 247, 13, 1, 1 )
+    my @bytes = encode_oid('{ pkcs-1 1 }');
+
+    # This registers a name with an OID.  It's saved internally and
+    # serves as repository of names for further parsing, such as 'pkcs-1'
+    # in the strings used above.
+    register_object('pkcs-1', '{ pkcs 1 }');
+
+
+    use OpenSSL::OID qw(:DEFAULT encode_oid_nums);
+
+    # This does the same as encode_oid(), but takes the output of
+    # parse_oid() as input.
+    my @bytes = encode_oid_nums(@nums);
+
+=head1 EXPORT
+
+The functions parse_oid and encode_oid are exported by default.
+The function encode_oid_nums() can be exported explicitly.
+
+=cut
+
+######## REGEXPS
+
+# ASN.1 object identifiers come in two forms: 1) the dotted form (referred
+# to as XMLObjIdentifierValue in X.690), 2) the bracketed form (referred to
+# as ObjectIdentifierValue in X.690)
```

Comment:
Can you swap these over so that the ObjectIdentifierValue comes before XMLObjectIdentifierValue. Both in this sentence and in the example. So that both the code below and the text here are in the same order?

Reply:
Sure

Reply:
Done